File: math/factorial/Haskell/factorial.hs
import System.IO

-- | The factorial function takes a non negative integer 'n' and calculates the 
-- product of all the positive integers before n
factorial :: Int -> Int
factorial n
    | n < 0     = error "factorial: Undefined for negative numbers"
    | n == 0    = 1
    | otherwise = n * factorial (n - 1)

main :: IO ()
main = do
    putStr "n = "
    -- Haskell buffers stdout until it is flushed or a new line is printed
    hFlush stdout
    n <- readLn :: IO Int
    putStrLn . show . factorial $ n

-- Tests
testBottomUp :: Int -> Bool
testBottomUp n = product [1..n] == factorial n
testTwenty :: Bool
testTwenty = factorial 20 == 2432902008176640000

File: math/factorial/js/Factorial.js
function factorial (n) {
  if (n == 0 || n == 1)
    return 1;
  if (f[n] > 0)
    return f[n];
  return f[n] = factorial(n-1) * n;
} 

//tests
factorial(5)
factorial(3)
factorial(5)
factorial(8)
factorial(10)